eng/pipelines/templates/steps/produce-net-standalone-packs.yml

```diff
@@ -17,48 +17,65 @@ parameters:
-    displayName: Create .NET standalone packs destination directory
-
   - ${{ each target in parameters.BuildMatrix }}:
     - pwsh: |
-        $destinationPathSegment = "$(Build.ArtifactStagingDirectory)/binaries/$((Split-Path -Leaf "${{ parameters.TargetDirectory }}"))-standalone-${{ target.rid }}"
-        $sourcePath = "${{ parameters.TargetDirectory }}/${{ target.rid }}"
-
-        Write-Host "dotnet publish -f ${{ target.framework }} -c Release -r ${{ target.rid }} --self-contained -p:PublishSingleFile=true --output ${{ parameters.TargetDirectory }}/${{ target.rid }}"
-        dotnet publish -f ${{ target.framework }} -c Release -r ${{ target.rid }} --self-contained -p:PublishSingleFile=true --output ${{ parameters.TargetDirectory }}/${{ target.rid }}
+        Write-Host "${{ and(not(startswith(target.rid, 'osx')), not(eq(variables['Agent.OS'], 'Darwin'))) }}"
+        Write-Host "${{ and(startsWith(target.rid, 'osx'), eq(variables['Agent.OS'], 'Darwin')) }}"
+        Write-Host "$(Agent.OS)"
+        Write-Host "${{ target.rid }}"
 
-        if ($LASTEXITCODE -ne 0) {
-          Write-Error "dotnet publish failed with exit code $LASTEXITCODE."
-          exit $LASTEXITCODE
-        }
+        Write-Host "Breakdown of mac expressions is as follows"
+        Write-Host "${{ startsWith(target.rid, 'osx') }}"
+        Write-Host "${{ eq(variables['Agent.OS'], 'Darwin') }}"
+        Write-Host "${{ variables['Agent.OS'] }}"
+      displayName: Dump Expression Outputs
 
-        # produce a tar.gz only for linux
-        if ("${{ target.rid }}".Contains("linux")){
-          # tar on powershell in linux has some weirdness. For instance, this is a proper call to tar when we don't want to include the relative path to the target folder
-          # tar -cvzf -C tools/test-proxy/linux-arm64 blah.tar.gz tools/test-proxy/linux-arm64 
-          # however when we use this, we actually get an error. To avoid this, we simply CD into the target directory before tar-ing it.
-          Push-Location "$sourcePath"
-          # The sum contents within this folder will be: `appSettings.json`, `test-proxy.pdb`, `test-proxy` (the binary), and a certificate.
-          # This statement grabs the first extensionless file within the produced binary folder, which will always be the binary we need to set the executable bit on.
-          $binaryFile = (Get-ChildItem -Path . | Where-Object { !([System.IO.Path]::HasExtension($_)) } | Select-Object -First 1).ToString().Replace("`\","/")
-          bash -c "chmod +x $binaryFile"
-          tar -cvzf "$($destinationPathSegment).tar.gz" .
-          Pop-Location
-        }
-        else {
-          Compress-Archive -Path "$($sourcePath)/*" -DestinationPath "$($destinationPathSegment).zip"
-        }
+    # ensure that we only build the mac standalone executables on a mac agent, everything else on not mac
+    - ${{ if startsWith(target.rid, 'osx') }}:
```

Comment:
Am I missing something or are these 2 code blocks exactly the same except for the condition on OS agent? Should we just validate OS agent and rid inside of the script?

Reply:
Well not necessarily validate but filter them to only run the osx rid entries on Mac.

Reply:
This is the only way to make that happen. `Agent.OS` is a _runtime_ variable, so cannot be used in template expression syntax. If you look at the actual steps that are executed, everything lines up.

Reply:
My original version of this was exactly a 


`if target.rid starts with osx and agent.os == darwin` -> run steps
`if target.rid doesn't start with osx and agent.os != darwin` -> run steps.

This combines it such that we only run the appropriate step on the right platform.

Reply:
I guess I was thinking we could stop looking in the template and do it in the script and just filter based on AgentOS and Rid. However, I agree this approach as well but we could do it dynamically.

Reply:
The artifact publish itself remains complex and wonky if we do it that way. I am confident this will work, and I have burned **far** too much time into the PR already.

I really appreciate you signed off while having this discussion, and we definitely can revisit. I'm certain that I will have another opportunity in here.